`TabPanel`: support manual tab activation

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

Closes #45390

## What?
<!-- In a few words, what is the PR actually doing? -->

Adds support for manual tab activation to the `TabPanel` component (see https://github.com/WordPress/gutenberg/issues/45390 and  [W3C docs](https://www.w3.org/WAI/ARIA/apg/patterns/tabpanel/) for more info)

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

New feature for the `TabPanel` component, for enhanced accessibility when necessary.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

- Introduced a new prop `hasManualTabActivation` (name TBD)
- The component keeps behaving as before when the new prop is not specified (backwards-compat)
- Added unit tests for both automatic and manual tab activation (including a necessary mock to test keyboard interactions)

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

- Load the default `TabPanel` example in Storybook
- Click on the first tab, then use arrow key to switch tab. Note how the currently shown tabpanel changes as the arrow keys are pressed
- Via Storybook controls, change the value of the `hasManualTabActivation` prop to `true`
- Click on the first tab, then use arrow keys to navigate through the list of tabs. Notice how focusing a tab doesn't cause the currently shown tab panel to change.
- Make sure that the tabpanel changes only when the space or enter keys are pressed while focusing a tab.

## Screenshots or screencast <!-- if applicable -->

https://user-images.githubusercontent.com/1083581/203551184-5f6f76d9-fa39-4128-bc72-dfd88d2b3f75.mp4

ciampo committed:
TabPanel: support manual tab activation

ciampo committed:
Mock `getClientRects` in TabPanel s unit tests to make sure that `NavigableMenu` works as expected

ciampo committed:
Add tab activation unit tests

ciampo committed:
Update docs

ciampo commented:
@jasmussen and @jameskoster , this may require design feedback as I don't think that `TabPanel` styles ever accounted for a tab to be focused while not being active.

ciampo committed:
CHANGELOG

ciampo commented on packages/components/src/tab-panel/README.md:
```diff
@@ -143,6 +143,14 @@ The name of the tab to be selected upon mounting of component. If this prop is n
 -   Required: No
 -   Default: none
 
+#### hasManualTabActivation
```
Not entirely convinced about this new prop's name, open to suggestions!

ntsekouras replied:
Maybe just `manualActivation`?

ciampo replied:
For boolean props I like to use names that start with `is` or `has` when possible (or at most, action-focused names), which makes it more obvious that the prop is a boolean — so maybe `hasManualActivation` or `activateTabsManually` ?

ntsekouras replied:
Your suggestion for `selectOnMove` sound good to me.

ciampo replied:
Done!

ntsekouras commented on packages/components/src/tab-panel/test/index.tsx:
```diff
@@ -34,7 +34,23 @@ const TABS = [
+		// the `focus.tabbable.find()` and by the `focus.focusable.find()` functions
+		// from the `@wordpress/dom` package.
+		// @ts-expect-error We're not trying to comply to the DOM spec, only mocking
+		window.HTMLElement.prototype.getClientRects = function () {
```
Curious why we need this.. I saw it's also the case for some other tests like `NavigableMenu`, but I'd love to understand better. Is it related with the testing library used?

ciampo replied:
`NavigableMenu` internally uses `focus.tabbable.find()` and `focus.focusable.find()` to search for the tabbable/focusable items — the logic requires a tabbable item [to be visible](https://github.com/WordPress/gutenberg/blob/feat/tab-panel-manual-tab-activation/packages/dom/build-module/focusable.js) — unfortunately, by default `js-dom` doesn't satisfy any of those checks, meaning that by default in our tests `NavigableMenu` won't move the selection to the next/previous tabs when interacting with the keyboard.

Mocking `getClientRects()` makes the above `isVisible` check pass and allows the test to work :)

ntsekouras replied:
So it's related to `js-dom`.. Thanks for elaborating Marco!

ciampo replied:
No worries! It took me a while to find out at first too 😅

ntsekouras reviewed (commented):
Thanks for the PR Marco! We just need to take a look at the non-active but focused styles, as you already mention. Other than that this looks good.

jameskoster commented:
It's working for me:

https://user-images.githubusercontent.com/846565/203558051-9e94a613-e57b-4909-a5d5-0213c3702315.mp4

I don't have a great idea for the name either, but perhaps something that relates more to the states involved; `openPanelOnFocus`?

ciampo commented:
> I don't have a great idea for the name either, but perhaps something that relates more to the states involved; `openPanelOnFocus`?

On a similar note, `ariakit` calls this feature `selectOnMove` — do we think it would be a better prop name?

jameskoster commented:
> @jasmussen and @jameskoster , this may require design feedback as I don't think that TabPanel styles ever accounted for a tab to be focused while not being active.

We can probably consider this when we get to https://github.com/WordPress/gutenberg/issues/41904.

chad1008 commented:
One thing that's occurring to me is that `ariakit` and this change both default to displaying the tab on focus, which is good, but we come at it from opposite directions. To enable manual selection:
- `ariakit` wants `selectOnMove` to be `false`
- This PR will want `hasManualTabActivation` (or whatever we name it) to be `true`

So we'd eventually end up managing the `ariakit` prop with something like `selectOnMove = { ! hasManualTabActivation }`

That's doable, of course, but maybe it'd be more readable in the long run if we went (as @ciampo suggests) with the same `selectOnMove` prop name, and flipped our boolean accordingly. That way our component and `ariakit`'s implementation are completely in sync.

alexstine reviewed (approved):
@ciampo Thanks for getting this done quickly. Hopefully, once this is merged, other PRs can progress and further enhancements can be made to code already in trunk.

Giving the PR an approval while the last few bits of feedback are addressed. I agree using the same prop name as we'll have in Ariakit could be useful.

tyxla commented on packages/components/src/tab-panel/index.tsx:
```diff
@@ -93,7 +94,12 @@ export function TabPanel( {
-	const onNavigate = ( _childIndex: number, child: HTMLButtonElement ) => {
+	// Simulate a click on the newly focused tab, which causes the component
+	// to show the `tab-panel` associated with the clicked tab.
+	const activeTabAutomatically = (
```
Nit: naming seems off here, did we mean to say `activateTabAutomatically`?

ciampo replied:
Yup! Done in https://github.com/WordPress/gutenberg/pull/46004/commits/8ef6ea4fe0c64949ad0a4a30d48a0560ae37d498

tyxla commented on packages/components/src/tab-panel/test/index.tsx:
```diff
@@ -34,7 +34,23 @@ const TABS = [
 
 const getSelectedTab = () => screen.getByRole( 'tab', { selected: true } );
 
+const originalGetClientRects = window.HTMLElement.prototype.getClientRects;
```
Nit: should this be declared right before we override it in `beforeAll`? This would then be a `let` and we'd just set it in `beforeAll()`.

ciampo replied:
Done in https://github.com/WordPress/gutenberg/pull/46004/commits/3128c2684b703a8636e6af1c5a29aad01cde82eb

tyxla commented on packages/components/src/tab-panel/test/index.tsx:
```diff
@@ -194,4 +210,98 @@ describe( 'TabPanel', () => {
+					tabs={ TABS }
+					children={ () => undefined }
+					onSelect={ mockOnSelect }
+					hasManualTabActivation={ true }
```
```suggestion
					hasManualTabActivation
```

ciampo replied:
No need for this anymore, since the new `selectOnMove` prop needs a `false` value here

tyxla reviewed (approved):
This is looking great, there's not much to add to what has been suggested in prior review already. 🚀

ciampo committed:
Rename prop to `selectOnMove`

ciampo committed:
Typo

ciampo committed:
Move `originalGetClientRects` declaration inside `beforeAll`

ciampo commented:
I [renamed the prop](https://github.com/WordPress/gutenberg/pull/46004/commits/ebcb84d1c7f4a3e6d9dfbc8f568e02807bc2aa30) to `selectOnMove` as agreed, and addressed all remaining feedback.

I guess the last thing to decide is whether we're ok to merge this PR with the current way tabs look in manual activation mode, when they are focused but not active (see video in the description).

tyxla commented on packages/components/src/tab-panel/README.md:
```diff
@@ -143,6 +143,14 @@ The name of the tab to be selected upon mounting of component. If this prop is n
 
+#### selectOnMove
+
+When `true`, the tab will selected when receiving focus (automatic tab activation). When `false`, the tab will be selected only when clicked (manual tab activation). See the [official W3C docs](https://www.w3.org/WAI/ARIA/apg/patterns/tabpanel/) for more info.
```
```suggestion
When `true`, the tab will be selected when receiving focus (automatic tab activation). When `false`, the tab will be selected only when clicked (manual tab activation). See the [official W3C docs](https://www.w3.org/WAI/ARIA/apg/patterns/tabpanel/) for more info.
```

tyxla commented on packages/components/src/tab-panel/types.ts:
```diff
@@ -70,4 +70,15 @@ export type TabPanelProps = {
 	 */
 	tabs: Tab[];
+	/**
+	 * When `true`, the tab will selected when receiving focus (automatic tab
```
```suggestion
	 * When `true`, the tab will be selected when receiving focus (automatic tab
```

tyxla reviewed (approved):
Looks great 🚀

ciampo committed:
Apply suggestions from code review

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

ciampo committed:
Merge branch 'trunk' into feat/tab-panel-manual-tab-activation

ciampo commented:
> whether we're ok to merge this PR with the current way tabs look in manual activation mode

Given the absence of strong feedback against merging, I'll go ahead and merge this PR once CI is ✅ . As @jameskoster mentioned above:

> We can probably consider this when we get to #41904.
